examples/binaries/reserve-gas/src/lib.rs

```diff
@@ -0,0 +1,87 @@
+// This file is part of Gear.
+
+// Copyright (C) 2022 Gear Technologies Inc.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use codec::{Decode, Encode};
+use gstd::{exec, msg, prelude::*, ReservationId};
+
+#[cfg(feature = "std")]
+mod code {
+    include!(concat!(env!("OUT_DIR"), "/wasm_binary.rs"));
+}
+
+#[cfg(feature = "std")]
+pub use code::WASM_BINARY_OPT as WASM_BINARY;
+
+static mut RESERVATION_ID: Option<ReservationId> = None;
+
+const RESERVATION_AMOUNT: u32 = 50_000_000;
+
+#[derive(Debug, Encode, Decode)]
+pub enum HandleAction {
+    Unreserve,
+    Exit,
+}
+
+#[no_mangle]
+unsafe extern "C" fn init() {
```

Comment:
Also case when program terminates should be covered by test!

Steps to reproduce:
1. Program starts `init()`
2. Reservation occurs (it's become valid and saved in storage)
3. `init()` goes to wait
4. program panics in `init()` after being waked somehow
5. assert!(reservation freed)